avr-hal-generic/src/adc.rs

```diff
@@ -0,0 +1,129 @@
+        pub use avr_hal::adc::*;
+
+        pub struct $Adc {
+            peripheral: $ADC,
+            is_reading: bool,
+        }
+
+        impl $Adc {
+            pub fn new(peripheral: $ADC, settings: AdcSettings) -> $Adc {
+                let s = Self { peripheral, is_reading: false } ;
+                s.enable(settings);
+                s
+            }
+
+            fn enable(&self, settings: AdcSettings) {
+                self.peripheral.adcsra.write(|w| {w.aden().set_bit();
+                                            match settings.adps {
+                                                ClockRateDivision::Factor2 => w.adps().val_0x01(),
+                                                ClockRateDivision::Factor4 => w.adps().val_0x02(),
+                                                ClockRateDivision::Factor8 => w.adps().val_0x03(),
+                                                ClockRateDivision::Factor16 => w.adps().val_0x04(),
+                                                ClockRateDivision::Factor32 => w.adps().val_0x05(),
+                                                ClockRateDivision::Factor64 => w.adps().val_0x06(),
+                                                ClockRateDivision::Factor128 => w.adps().val_0x07(),
+                                            }});
+                self.peripheral.admux.write(|w| match settings.aref {
+                    ReferenceVoltage::Aref => w.refs().val_0x00(),
+                    ReferenceVoltage::Vcc => w.refs().val_0x01(),
+                    ReferenceVoltage::Internal => w.refs().val_0x03(),
+                });
+                                                
+            }
+
+            fn disable(&self) {
+                self.peripheral.adcsra.reset();
+            }
+
+            pub fn release(self) -> $ADC {
+                self.disable();
+                self.peripheral
+            }
+        }
+
+        impl<WORD, PIN> OneShot<$Adc, WORD, PIN> for $Adc
+        where
+            WORD: From<u16>,
+            PIN: Channel<$Adc, ID=$ID>,
+        {
+            type Error = ();
+
+            fn read(&mut self, _pin: &mut PIN) -> nb::Result<WORD, Self::Error> {
+                match (self.is_reading, self.peripheral.adcsra.read().adsc().bit_is_set()) {
+                    (true, true) =>  Err(nb::Error::WouldBlock),
+                    (true, false) => {
+                        self.is_reading = false;
+                        Ok(self.peripheral.adc.read().bits().into())
+                    },
+                    (false, _) => {
+                        self.is_reading = true; self.peripheral.admux.modify(|_, w| w.adlar().clear_bit()
+                                                                                     .mux().variant(PIN::channel()));
+                        self.peripheral.adcsra.modify(|_, w| w.adsc().set_bit());
+                        Err(nb::Error::WouldBlock)
+                    }
+                }
+            }
```

Comment:
I think this method has a problem:

If I call it once with a pin 1, it will return `WouldBlock`.  If I then poll with a different pin 2, until I receive a result, I will get a result for pin 1.  Not sure what upstream `embedded-hal` thinks of this; the docs don't mention this case ... I would suggest to restart the measurement and start sampling pin 2 instead. Or take a look what other implementations are doing here ...

Reply:
Changed the implementation so every pin has its own `is_reading`flag, I think that should work

Reply:
Yes, this would work but I feel like 8 `bool`s just to work around a corner case is a bit much to pay.  Can't we reuse the value of `ADMUX[MUX]` here?  I think it should be possible to compare it against the enum variant for the pin (`PIN::channel()`) and base the decision off of that.

Reply:
Right a single field with an enum would be better.
I would like to use ADMUX[MUX] but it doesn't have a state for no Channel selected which is my initial state of the Adc,

Reply:
Fair point, the new implementation looks good!